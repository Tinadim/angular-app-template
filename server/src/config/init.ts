import 'zone.js/dist/zone-node';
import { resolve } from 'path';
import * as appRoot from 'app-root-path';
import * as dotenv from 'dotenv';
import { registerProvider } from '@app/client-provider';
import * as provider from '@client/src/client-provider';

dotenv.config({ path: resolve(appRoot.path, '.env') });

console.log('<<<<< ROOT', appRoot.path, '>>>>>');

const resolvedPath = require.resolve('../../../client/src/client-provider');
// const resolvedPath = resolve(__dirname, '../../client/src/client-provider');

console.log('ENV\n\n', JSON.stringify(process.env.USE_SSR));

registerProvider(resolvedPath, provider);
