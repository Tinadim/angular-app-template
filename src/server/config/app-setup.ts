import express, { Application, Request, Response } from 'express';
import bodyParser from 'body-parser';
import Router from 'express-promise-router';
import { ngExpressEngine } from '@nguniversal/express-engine';
import { APP_BASE_HREF } from '@angular/common';
import { resolve } from 'path';
import { mount } from '../api';
import { AppServerModule } from '../server-side-rendering/app-server-module';

const distFolder = resolve(__dirname, '../../../dist/client');
const useSsr = process.env.USE_SSR;
const renderer = useSsr
    ? (req: Request, res: Response) => res.render('index', { req, providers: [{ provide: APP_BASE_HREF, useValue: req.baseUrl }] })
    : (req: Request, res: Response) => res.sendFile(`${distFolder}/index.html`);

export function buildApp() {
    const app = express();
    setupMiddleware(app);
    bootEndpoints(app);
    serveStaticFiles(app);
    return app;
}

function serveStaticFiles(app: Application) {
    app.get('*.*', express.static(distFolder, {
        maxAge: '1y'
    }));

    app.get('*', renderer);
}

function setupMiddleware(app: Application) {
    if (useSsr) {
        app.engine('html', ngExpressEngine({
            bootstrap: AppServerModule,
        }));
    }

    app.set('view engine', 'html');
    app.set('views', distFolder);
    app.use(bodyParser.urlencoded({ extended: true }));
    app.use(bodyParser.json());
}

function bootEndpoints(app: Application) {
    const router = Router();
    mount(router);
    app.use('/api', router);
}
