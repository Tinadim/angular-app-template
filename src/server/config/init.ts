import 'zone.js/dist/zone-node';
import { resolve } from 'path';
import { config as configEnv } from 'dotenv-safe';

const example = resolve(process.cwd(), '.env.example');
configEnv({ example });

console.log('<<<<< USE_SSR', process.env.USE_SSR, '>>>>>');
