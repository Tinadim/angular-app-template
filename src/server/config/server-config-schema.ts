import { $enum } from 'ts-enum-util';

enum EnvVars {
    NodeEnv = 'NODE_ENV',
    Port = 'PORT',
    UseShutDownHandler = 'USE_SHUTDOWN_HANDLER',
    UseSsr = 'USE_SSR'
}

enum ArgVars {
    Port = 'port',
    UseShutDownHandler = 'useShutdownHandler',
    UseSsr = 'useSsr'
}

export const ENV_VARS = $enum(EnvVars).getValues();

export const ARG_VARS = $enum(ArgVars).getValues();

export const DEFAULTS = {
    port: 8080,
    useSsr: true,
    useShutdownHandler: false
};

export const envSchema: object = {
    properties: {
        [EnvVars.Port]: {
            type: 'number',
            default: DEFAULTS.port,
        },
        [EnvVars.UseShutDownHandler]: {
            type: 'boolean',
            default: DEFAULTS.useShutdownHandler,
        },
        [EnvVars.UseSsr]: {
            type: 'boolean',
            default: DEFAULTS.useSsr,
        }
    },
    type: 'object'
};

export const argvSchema: object = {
    properties: {
        [ArgVars.Port]: {
            type: 'number',
        },
        [ArgVars.UseShutDownHandler]: {
            type: 'boolean',
        },
        [ArgVars.UseSsr]: {
            type: 'boolean',
        }
    },
    type: 'object'
};
