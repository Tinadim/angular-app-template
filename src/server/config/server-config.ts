import Ajv from 'ajv';
import { camelCase, cloneDeep, mapKeys, pick } from 'lodash';
import { resolve } from 'path';
import { parse as parseArgv } from 'yargs';
import { envSchema, argvSchema, ENV_VARS, ARG_VARS } from './server-config-schema';

export class ServerConfig {
    readonly distFolderPath: string;
    readonly port: number;
    readonly isProduction: boolean;
    readonly useShutdownHandler: boolean;
    readonly useSsr: boolean;

    constructor() {
        this.distFolderPath = resolve(process.cwd(), './dist/client');

        const mergedConfigs = this.mergeConfigs();

        this.isProduction = process.env.NODE_ENV === 'production';
        this.port = mergedConfigs.port;
        this.useSsr = mergedConfigs.useSsr;
        this.useShutdownHandler = mergedConfigs.useShutdownHandler;
    }

    private mergeConfigs() {
        const ajv = new Ajv({
            allErrors: true,
            coerceTypes: true,
            format: 'full',
            useDefaults: true,
        });
        const relevantEnvVars = pick(process.env, ENV_VARS);
        const argvCopy = pick(cloneDeep(parseArgv()), ARG_VARS);
        this.validateEnv(ajv, relevantEnvVars);
        this.validateArgv(ajv, argvCopy);

        const mappedEnvVars = mapKeys(relevantEnvVars, (value: any, key: string) => camelCase(key));
        return Object.assign({}, mappedEnvVars, argvCopy) as any;
    }

    private validateEnv(ajv: Ajv.Ajv, envVars: object) {
        const valid = ajv.validate(envSchema, envVars);
        if (!valid) {
            throw this.formatAjvError(ajv.errors);
        }
    }

    private validateArgv(ajv: Ajv.Ajv, argVars: object) {
        const valid = ajv.validate(argvSchema, argVars);
        if (!valid) {
            throw this.formatAjvError(ajv.errors);
        }
    }

    private formatAjvError(errors: Ajv.ErrorObject[]) {
        // TODO add better error formatting
        throw new Ajv.ValidationError(errors);
    }
}

export const serverConfig = new ServerConfig();
