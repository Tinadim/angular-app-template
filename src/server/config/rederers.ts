import { Request, RequestHandler, Response } from 'express';
import { APP_BASE_HREF } from '@angular/common';
import { serverConfig } from './server-config';

const { useSsr, distFolderPath } = serverConfig;

export function renderIndex(req: Request, res: Response) {
    res.render('index', {
        req,
        providers: [{ provide: APP_BASE_HREF, useValue: req.baseUrl }],
    });
}

export function sendIndex(req: Request, res: Response) {
    res.sendFile(`${distFolderPath}/index.html`);
}

export function getRenderer(): RequestHandler {
    return useSsr ? renderIndex : sendIndex;
}
