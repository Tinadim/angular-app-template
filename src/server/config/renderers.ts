import { Request, RequestHandler, Response } from 'express';
import { APP_BASE_HREF } from '@angular/common';
import { serverConfig } from './server-config';

export function renderIndex(req: Request, res: Response) {
    res.render('index', {
        req,
        providers: [{ provide: APP_BASE_HREF, useValue: req.baseUrl }],
    });
}

export function sendIndex(req: Request, res: Response) {
    const { distFolderPath } = serverConfig;
    res.sendFile(`${distFolderPath}/index.html`);
}

export function getRenderer(): RequestHandler {
    return serverConfig.useSsr ? renderIndex : sendIndex;
}
