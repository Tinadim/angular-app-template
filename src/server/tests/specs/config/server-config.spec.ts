import { concat } from 'lodash';
import { ServerConfig } from '@server/config/server-config';

jest.mock('yargs', () => {
    const actualImplementation = jest.requireActual('yargs');
    return {
        parse: () => {
            return actualImplementation.parse(process.argv.splice(2));
        }
    };
});

describe('server-config.spec.ts', () => {
    const envBackup = process.env;
    const argvBackup = process.argv;

    beforeEach(() => {
        jest.resetModules();
        process.argv = [...argvBackup];
        process.env = { ...envBackup };
    });

    afterAll(() => {
        process.argv = argvBackup;
        process.env = envBackup;
    });

    it('Should set isProduction to true if NODE_ENV is set to production', () => {
        process.env.NODE_ENV = 'production';
        const config = new ServerConfig();
        expect(config.isProduction).toBe(true);
    });

    it('Should set isProduction to false if NODE_ENV is set to something else', () => {
        process.env.NODE_ENV = 'dev';
        const config = new ServerConfig();
        expect(config.isProduction).toBe(false);
    });

    it('Should use configs from env vars', () => {
        process.env.PORT = '8181';
        process.env.USE_SSR = 'false';
        process.env.USE_SHUTDOWN_HANDLER = 'true';

        const config = new ServerConfig();

        expect(config.port).toBe(8181);
        expect(config.useShutdownHandler).toBe(true);
        expect(config.useSsr).toBe(false);
    });

    it('Should use configs from argv', () => {
        process.argv = concat(process.argv, [
            '--port',
            '8282',
            '--useShutdownHandler',
            'true'
        ]);

        const config = new ServerConfig();

        expect(config.port).toBe(8282);
        expect(config.useShutdownHandler).toBe(true);
    });

    it('Should use default configs if nothing is provided', () => {
        const config = new ServerConfig();

        expect(config.port).toBe(8080);
        expect(config.useShutdownHandler).toBe(false);
        expect(config.useSsr).toBe(true);
    });

    it('Should merge configs from env and argv', () => {
        process.env.PORT = '8181';
        process.env.USE_SSR = 'false';
        process.env.USE_SHUTDOWN_HANDLER = 'true';

        process.argv = concat(process.argv, [
            '--port',
            '8282',
        ]);

        const config = new ServerConfig();

        expect(config.port).toBe(8282);
        expect(config.useShutdownHandler).toBe(true);
        expect(config.useSsr).toBe(false);
    });

    it('Should throw an error if invalid env vars are provided', () => {
        process.env.PORT = 'not-a-number';
        process.env.USE_SSR = 'not-a-boolean';
        process.env.USE_SHUTDOWN_HANDLER = 'not-a-boolean';

        const run = () => new ServerConfig();

        expect(run).toThrowErrorMatchingSnapshot();
    });

    it('Should throw an error if invalid arg vars are provided', () => {
        process.argv = concat(process.argv, [
            '--port',
            'not-a-number',
            '--useSsr',
            'not-a-boolean',
            '--useShutdownHandler',
            'not-a-boolean'
        ]);

        const run = () => new ServerConfig();

        expect(run).toThrowErrorMatchingSnapshot();
    });
});
