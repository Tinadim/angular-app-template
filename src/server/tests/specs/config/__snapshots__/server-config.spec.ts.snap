// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`server-config.spec.ts Should throw an error if invalid arg vars are provided 1`] = `"validation failed"`;

exports[`server-config.spec.ts Should throw an error if invalid env vars are provided 1`] = `"validation failed"`;
